### PR TITLE
Clean up abandoned WAL directories

### LINKED ISSUE #132: Scraping service: WAL management
If a Prometheus instance is deleted, there's no reason to keep the WAL around any more. It should be removed from disk after the instance is shut down. The implementation of this should be careful to make sure this only happens if the instance is explicitly deleted; restarts of the instance or the Agent process should keep the WAL.

It's possible that the Agent could restart and not be reassigned any of the instances it had before. This would leave the old WALs sitting around on disk; there should be a really slow poll interval to check for old WALs and delete them.

The WAL should also be given a max age. The truncation loop will use the higher value from last succesful send and the max age to truncate old data. This ensures that the WAL does not grow forever. The default value here should be really generous, 12 hours is a good start.

### PR DESCRIPTION
**Pull request by 56quarters**
Remove WAL directories that are no longer associated with a `ManagedInstance`
that this agent is responsible for. The storage directories used by active
instances are compared to all subdirectories under the `wal_directory` root.
Any that are not associated with an instance and that haven't been modified
in over a configured amount of time are removed.

How often the check is run and the cutoff for how recently a WAL must have
been modified are controlled by the `wal_cleanup_period` and
`wal_cleanup_age` settings respectively, under the top-level Prometheus
configuration. When omitted, they default to a period of 30 minutes and
recently modified threshold of 12 hours.

Fixes #132

- [x] CHANGELOG updated 
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by 56quarters** [comment]
@rfratto I'm not really sure about the best way to integrate this into the rest of the agent. My initial thought was that a `Cleaner` instance would be a member of the `Agent` struct. The goroutine to run it periodically would be started on construction of the `Agent` struct. Your thoughts?

**Comment by rfratto** [comment]
Thanks for working on this! 

I think creating it from the Agent struct is fine, but I think I'd prefer the goroutine to be launched by `NewCleaner` with a `Stop` method that signals the goroutine to exit. I'd like to keep the Agent acting as just the glue for other pieces of logic and not have any real logic itself. We do something similar in `newManager`, where it launches its own goroutine and has a `Stop` method to close it.

File: pkg/prom/wal/util.go
Comment (56quarters) [inline pkg/prom/wal/util.go:116]: Hmm, maybe this should be `filepath.Join`?

Comment (56quarters) [reply pkg/prom/wal/util.go:116]: Yep, duh. That's what the code it replaced did.

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:38]: Can you rename the name of counters to end in `_total`? I'd like to stick to the defined Prometheus [best practices](https://prometheus.io/docs/practices/naming/) when possible.

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:222]: Can we add a histogram to track the time it takes to perform a clean cycle? I'm interested in how long it will take for significant deployments of the Agent.

File: pkg/prom/cleaner_test.go
Comment (rfratto) [inline pkg/prom/cleaner_test.go:19]: nit: I'd rather not define an extra function here. It adds more to think about when reading the code ("what kind of logger does getLogger return?"). To avoid the repetition, you can just define the logger at the top of the tests:

```go
logger := log.NewLogfmtLogger(os.Stderr)
cleaner := NewWALCleaner(logger, &mockManager{}, walRoot)
```

(I can be convinced otherwise though)

File: pkg/prom/cleaner_test.go
Comment (rfratto) [inline pkg/prom/cleaner_test.go:161]: Can you use `instances.MockManager` here so we don't define these in too many places?

File: docs/configuration-reference.md
Comment (rfratto) [inline docs/configuration-reference.md:161]: Sorry for adding this feedback on the docs, but can we make the WAL cleanup optional? If the cleanup period is configured to `0s`, let's disable it.

File: pkg/prom/cleaner_test.go
Comment (rfratto) [inline pkg/prom/cleaner_test.go:28]: It's a little odd to see `require` and `assert` mixed together; can we consistently use `require` throughout the tests?

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:69]: Can we wrap these errors to make it more discoverable about which line is failing?

```go
return empty, fmt.Errorf("could not open WAL: %w", err)
```

(ditto for the rest of this function)

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:60]: IMO we don't need this to be an interface and can just define it as a function type:

```go
type lastModifiedFunc func(path string) (time.Time, error) 

func walLastModified(path string) (time.Time, error) {
  // ...
}
```

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:65]: nit: I'd prefer not defining the zero value for `time.Time` as a variable so it doesn't have to be repeated. Seeing a variable makes me want to look for all the places it's used. Can we just `return time.Time{}, err` directly in the failure cases? 

(Also, it's slowly becoming a standard to define variables with the zero value in the form `var empty time.Time` rather than using the `:=` syntax)

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:183]: Can we flatten this out into one if tree?

```go
if os.IsNotExist(err) { 
   // ...
} else if err != nil {
  // .. 
} else if info.IsDir() && filepath.Dir(p) == c.walDirectory {
  // ... 
}
return nil 
```

If you feel it looks better you can try a switch too:

```go
switch {
case os.IsNotExist(err):
case err != nil:
case info.IsDir() && filepath.Dir(p) == c.walDirectory:
}
```

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:196]: I'd prefer doing an early continue here to not have too much indented code:

```go
if managed[dir] {
  level.Debug(c.logger).Log("msg", "active WAL", "name", dir)
  continue 
}
```

File: pkg/prom/cleaner.go
Comment (rfratto) [inline pkg/prom/cleaner.go:213]: Shouldn't this be logged in the `if diff > c.minAge` block?

**Review by rfratto — COMMENTED** [review]
Thanks for working on this, this will be great for people using the scraping service! 🎉 This is pretty close to a mergeable state, though I think there's some work that we can do to make the code a bit smaller.

Aside from my smaller pieces of feedback, I have one larger one. Some linters require function and type comments to be commented in the "godoc" style which are strings optionally starting with A/An/The and the name of the thing being commented:

```go
// The Cleaner is responsible for ...
type Cleaner struct {}

// WithCleanerMinAge overrides the default age after which ....
func WithCleanerMinAge(d time.Duration) WALCleanerOpts 
```

We don't have that linting rule enabled (apparently, though I thought we did) and it only applies the restriction to exported functions/types, but I personally like doing it consistently for all types and functions. Can you make sure that at least all exported types and functions you added are written in this style?

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:128]: Yep, I can change that.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:114]: Whoops, yeah.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:222]: Good idea, will do.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:236]: Sounds good.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:248]: Good idea

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:60]: Ok, I can change that.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:65]: Ok, will do.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:183]: Yep. Any preference on if/else vs switch?

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:196]: Sure, will do.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:213]: Maybe? I put it outside the block explicitly since I wanted to be able to tell when a WAL was abandoned but not for long enough. I can move it since that's not really needed anymore for my testing or I can add another one to the `else` case. Thoughts?

Comment (rfratto) [reply pkg/prom/cleaner.go:213]: ah, I see now. At first I thought it was supposed to log unused WALs that will be deleted, not all unused WALs. This is fine as-is 👍

File: docs/configuration-reference.md
Comment (56quarters) [reply docs/configuration-reference.md:161]: Sure. I was thinking the same thing, that this should probably have an off switch

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:38]: Sure, will do.

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:69]: Sure

File: pkg/prom/cleaner.go
Comment (56quarters) [reply pkg/prom/cleaner.go:222]: Good idea

File: pkg/prom/cleaner_test.go
Comment (56quarters) [reply pkg/prom/cleaner_test.go:19]: Sounds good

File: pkg/prom/cleaner_test.go
Comment (56quarters) [reply pkg/prom/cleaner_test.go:28]: Whoops, will fix.

File: pkg/prom/cleaner_test.go
Comment (56quarters) [reply pkg/prom/cleaner_test.go:161]: Whoops, yep.

**Review by rfratto — APPROVED** [review]
LGTM, thank you! Once you rebase and fix the merge conflict, feel free to merge.

**Pull request merged**
